architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+| **In-place Upgrade**     | An upgrade approach that updates an existing Radius installation in its current location without requiring complete reinstallation. The process preserves all user data, configurations, and deployed application metadata while minimizing downtime of the control plane. This contrasts with alternative approaches like parallel deployments or complete reinstallations. |
+| **Rolling Upgrade**      | An upgrade strategy that updates components incrementally (one by one) rather than all at once, reducing downtime.                                                                                                                                                                                                                                                           |
+| **User Data Backup**     | A point-in-time backup of user data taken before an upgrade to enable recovery if problems occur.                                                                                                                                                                                                                                                                            |
+| **User Data Restore**    | The process of reverting user data to a previously backed up state after an upgrade attempt encounters issues.                                                                                                                                                                                                                                                               |
+| **Pre-flight Checks**    | Validation steps performed before an upgrade to ensure prerequisites are met and the system is in a valid state for upgrading.                                                                                                                                                                                                                                               |
+| **Version Skipping**     | The ability to upgrade directly from one version to a newer non-adjacent version (e.g., v0.40 → v0.44) without installing intermediate versions.                                                                                                                                                                                                                             |
+
+## Assumptions
+
+1. **User permissions**: Users running the upgrade command have enough permissions on both the Kubernetes cluster and the Radius installation.
+
+2. **Resource requirements**: The Kubernetes cluster has sufficient compute resources (CPU, memory) to run both the existing and new version components during the rolling upgrade process.
```

Comment:
What's the expected user experience during upgrade if resources are constrained?

Reply:
Good point. Users may hit mid-upgrade failures with no clear information. To fix this, we can:
- Update the Assumptions section (in docs or somewhere that the users can see) to call out about the minimum requirements for the `rad upgrade kubernetes` process
- Add ResourceAvailability pre-flight check that can measure the resources in the cluster and abort the process if there is not enough
- Improve the CLI UX so that on failure it can mention that there is not enough resources

We can apply all these steps and I will update the doc accordingly.